wallet-new/src/Cardano/Wallet/Kernel/Pending.hs

```diff
@@ -8,57 +8,112 @@ module Cardano.Wallet.Kernel.Pending (
-            return $ Right ()
+--
+-- NOTE: we select "our" output addresses from the transaction and pass it along to the data layer
+newPending :: ActiveWallet
+           -> HdAccountId
+           -> TxAux
+           -> Maybe TxMeta
+           -> IO (Either NewPendingError ())
+newPending w accountId tx mbMeta = do
+    newTx w accountId tx mbMeta $ \ourAddrs ->
+        update' ((walletPassive w) ^. wallets) $ NewPending accountId (InDb tx) ourAddrs
 
 -- | Submit new foreign transaction
 --
 -- A foreign transaction is a transaction that transfers funds from /another/
 -- wallet to this one.
-newForeign :: ActiveWallet -> HdAccountId -> TxAux -> TxMeta -> IO (Either NewForeignError ())
-newForeign ActiveWallet{..} accountId tx meta = do
-    res <- update' (walletPassive ^. wallets) $ NewForeign accountId (InDb tx)
+newForeign :: ActiveWallet
+           -> HdAccountId
+           -> TxAux
+           -> TxMeta
+           -> IO (Either NewForeignError ())
+newForeign w accountId tx meta = do
+    newTx w accountId tx (Just meta) $ \ourAddrs ->
+        update' ((walletPassive w) ^. wallets) $ NewForeign accountId (InDb tx) ourAddrs
+
+-- | Submit a new transaction
+--
+-- Will fail if the HdAccountId does not exist or if some inputs of the
+-- new transaction are not available for spending.
+--
+-- If the transaction is successfully added to the wallet state, transaction metadata
+-- is persisted and the submission layer is notified accordingly.
+--
+-- NOTE: we select "our" output addresses from the transaction and pass it along to the data layer
+newTx :: forall e. ActiveWallet
+      -> HdAccountId
+      -> TxAux
+      -> Maybe TxMeta
+      -> ([AddrWithId] -> IO (Either e ())) -- ^ the update to run, takes ourAddrs as arg
+      -> IO (Either e ())
+newTx ActiveWallet{..} accountId tx mbMeta upd = do
+    -- run the update
+    allOurs' <- allOurs <$> getWalletCredentials walletPassive
+    res <- upd allOurs'
     case res of
-        Left e -> return (Left e)
+        Left e   -> return (Left e)
         Right () -> do
-            putTxMeta (walletPassive ^. walletMeta) meta
-            modifyMVar_ walletSubmission (return . addPending accountId (Pending.singleton tx))
-            return $ Right ()
+            -- process transaction on success
+            putTxMeta' mbMeta
+            submitTx accountId tx
+            return (Right ())
+    where
+        addrs = NE.toList $ map txOutAddress (_txOutputs . taTx $ tx)
+        wid   = WalletIdHdRnd (accountId ^. hdAccountIdParent)
+
+        -- | NOTE: we recognise addresses in the transaction outputs that belong to _all_ wallets, 
+        --  not only for the wallet to which this transaction is being submitted
+        allOurs :: [(WalletId, EncryptedSecretKey)] -> [AddrWithId]
+        allOurs = concatMap (ourAddrs . snd)
+
+        ourAddrs :: EncryptedSecretKey -> [AddrWithId]
+        ourAddrs esk' =
+            map swap $ prefilter wKey identity addrs
```

Comment:
Can we rename this `prefilter` function? It's got nothing to do with prefiltering.